### ISSUE TITLE
[FEA] Delete conversion to `column_view` for r-value reference of `column` and `*column_wrapper`

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#11123]
**Is your feature request related to a problem? Please describe.**

It is a common error to accidentally pass a temporary `column` or `column_wrapper` object into a function taking a `column_view`.

This leads to dangling references as the `column_view` object will not keep the temporary `column(_wrapper)` object alive. 

**Describe the solution you'd like**

Delete the conversion operator for r-value refs.

```
struct column_view{};

struct column_wrapper{
    operator column_view() const& {
        return {};
    }
    operator column_view() && = delete;
};

void foo(column_view){}

int main(){
    column_wrapper w;
    foo(w);
    // foo(column_wrapper{}); // fails! This is what you want to happen
}
```

https://godbolt.org/z/en6a3E7eE

**Comment by jrhemstad** [comment]
For posterity:

There's two subtly different concepts at play here.

So we know that the lifetime of a temporary can be extended by binding it to a reference:
```std::string const& ref = std::string{"temporary"};```
The temporary string created is bound to the lifetime of `ref`. Simple stuff.

Extending this to calling a function like:
```
void foo_lref( std::string const& ref );
void foo_rref( std::string&& rref);
void foo_value( std::string v);

foo_lref(std::string{"temporary"});
foo_rref(std::string{"temporary"});
foo_value( std::string{"temporary"});
```

All of these calls are okay. The first two the temporary binds to the reference (or so I thought!), the third invokes the `string(string&&)` ctor and constructs a new object from the temporary.

Here's the subtly different thing:
```
void foo_stringview(std::string_view);

foo_stringview(std::string{"temporary"});
```

As we've seen this is also okay. But _not_ because the temporary is binding to any reference. It's as you quoted before, I just didn't grok the importance of it until now:
> All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created

`foo_stringview(std::string{"temporary"});` is an expression and the temporary string is not destroyed until the end of the expression (when the function returns).

So my confusion originated from assuming that passing temporaries to functions worked by the same mechanism as binding temporaries to references. But that isn't really the case.

In other words, the reason that this works:
```
void foo(std::string const& s);
foo( std::string{"temporary"} );
```

Is different from the reason of why this works:
```
std::string const& s = std::string{"temporary"};
```

**Comment by bdice** [comment]
> It's as you quoted before:

This was in reference to my previous comment quoting [an exception to the lifetime of a temporary](https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary):
> a temporary bound to a reference parameter in a function call exists until the end of the full expression containing that function call: if the function returns a reference, which outlives the full expression, it becomes a dangling reference.

The same lifetime extension logic applies from `string_view` function parameters accepting `string&&` to `column_view` function parameters accepting `column&&`: https://stackoverflow.com/questions/46736467/why-doesnt-string-view-disallow-construction-from-string-temporaries

@jrhemstad also provided this nice Compiler Explorer demo: https://godbolt.org/z/MKvW3Mn3r
